connection_pool/connection_pool.go

```diff
@@ -0,0 +1,657 @@
+}
+
+// DeleteAsync sends deletion action to tarantool and returns Future.
+// Future's result will contain array with deleted tuple.
+func (connPool *ConnectionPool) DeleteAsync(space, index interface{}, key interface{}) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(RW)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.DeleteAsync(space, index, key), nil
+}
+
+// UpdateAsync sends deletion of a tuple by key and returns Future.
+// Future's result will contain array with updated tuple.
+func (connPool *ConnectionPool) UpdateAsync(space, index interface{}, key, ops interface{}) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(RW)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.UpdateAsync(space, index, key, ops), nil
+}
+
+// UpsertAsync sends "update or insert" action to tarantool and returns Future.
+// Future's sesult will not contain any tuple.
+func (connPool *ConnectionPool) UpsertAsync(space interface{}, tuple interface{}, ops interface{}) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(RW)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.UpsertAsync(space, tuple, ops), nil
+}
+
+// CallAsync sends a call to registered tarantool function and returns Future.
+// It uses request code for tarantool 1.6, so future's result is always array of arrays
+func (connPool *ConnectionPool) CallAsync(functionName string, args interface{}, userMode Mode) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(userMode)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.CallAsync(functionName, args), nil
+}
+
+// Call17Async sends a call to registered tarantool function and returns Future.
+// It uses request code for tarantool 1.7, so future's result will not be converted
+// (though, keep in mind, result is always array)
+func (connPool *ConnectionPool) Call17Async(functionName string, args interface{}, userMode Mode) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(userMode)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.Call17Async(functionName, args), nil
+}
+
+// EvalAsync sends a lua expression for evaluation and returns Future.
+func (connPool *ConnectionPool) EvalAsync(expr string, args interface{}, userMode Mode) (*tarantool.Future, error) {
+	conn, err := connPool.getNextConnection(userMode)
+	if err != nil || conn == nil {
+		return nil, err
+	}
+
+	return conn.EvalAsync(expr, args), nil
+}
+
+//
+// private
+//
+
+func (connPool *ConnectionPool) getConnectionRole(conn *tarantool.Connection) (Role, error) {
+	resp, err := conn.Call17("box.info", []interface{}{})
+	if err != nil {
+		return unknown, err
+	}
+	if resp == nil {
+		return unknown, ErrIncorrectResponse
+	}
+	if len(resp.Data) < 1 {
+		return unknown, ErrIncorrectResponse
+	}
+
+	replicaRole, ok := resp.Data[0].(map[interface{}]interface{})["ro"]
```

Comment:
Last Pythoon master discovery revision checks if `box.info().status == 'running'` for instance health

Reply:
I forgot about it, fixed, thanks.